lightning/src/ln/channelmanager.rs

```diff
@@ -1528,29 +1530,71 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 	/// Note that ALL inputs in the transaction pointed to by funding_txo MUST spend SegWit outputs
 	/// or your counterparty can steal your funds!
 	///
+	/// Returns an [`APIError::APIMisuseError`] if the funding_transaction spent non-SegWit outputs
+	/// or if no output was found which matches the parameters in [`Event::FundingGenerationReady`].
+	///
 	/// Panics if a funding transaction has already been provided for this channel.
 	///
-	/// May panic if the funding_txo is duplicative with some other channel (note that this should
-	/// be trivially prevented by using unique funding transaction keys per-channel).
-	pub fn funding_transaction_generated(&self, temporary_channel_id: &[u8; 32], funding_txo: OutPoint) {
+	/// May panic if the output found in the funding transaction is duplicative with some other
+	/// channel (note that this should be trivially prevented by using unique funding transaction
+	/// keys per-channel).
+	///
+	/// Do NOT broadcast the funding transaction yourself. When we have safely received our
+	/// counterparty's signature the funding transaction will automatically be broadcast via the
+	/// [`BroadcasterInterface`] provided when this `ChannelManager` was constructed.
+	pub fn funding_transaction_generated(&self, temporary_channel_id: &[u8; 32], funding_transaction: Transaction) -> Result<(), APIError> {
 		let _persistence_guard = PersistenceNotifierGuard::new(&self.total_consistency_lock, &self.persistence_notifier);
 
+		for inp in funding_transaction.input.iter() {
+			if inp.witness.is_empty() {
+				return Err(APIError::APIMisuseError {
+					err: "Funding transaction must be fully signed and spend Segwit outputs".to_owned()
```

Comment:
By transaction must be fully signed you meant use only SIGHASH_ALL otherwise you might have in-flight malleability breaking the signatures for the first commitment transaction ? 

Ideally we have the signatures available in the witnesses so in the future we might parse them to verify sighash types flag to protect users against footgunish behaviors.

Reply:
Hmm, we do have them here, but we'd need to parse the original script_pubkey, which I don't think we want users to have to include. I can include it in the docs.

Reply:
If the witness change in the wrong way with a SIGHASH_SINGLE as you point it that's not fine. Should we document "don't construct the transaction such that someone can arbitrarily modify it " ? I would say we suffer from cognitive dissonance there. It's quite obvious to us but in practice users might authorize malleation or even RBF for fee management (see https://lists.linuxfoundation.org/pipermail/lightning-dev/2021-March/002981.html).

Reply:
Right, I guess we could re-enforce the last line - if you don't broadcast the transaction yourself, you can't RBF either :).